### PR TITLE
raft store: follower replication for write-flow

### LINKED ISSUE #13269: raftstore: follower replication for write-flow
## Develop Task
1. A new mechanism in Raft algorithm which allows a follower to replicate log entries or snapshots to other followers and learners. This feature is mainly aimed to reduce cross-AZ (short for Availability Zones) traffic on cloud. RFC: https://github.com/tikv/rfcs/pull/98
1.1. Definition of new message type `MsgGroupBroadcast` (in raft-rs).
1.2. Handler for new message type  (in raft-rs).
1.3. Merge `MsgAppend` and replace it with `MsgGroupBroadcast` when peer sends raft messages.
3. Maintaining AZ information in peer from PD.

### PR DESCRIPTION
**Pull request by LintianShi**
### What is changed and how it works?

Issue Number: Close #13269

What's Changed:

```commit-message
1. Record and maintain availability zone information in `PollContext`.
2. Merge MsgAppend into MsgGroupBroadcast in `fn build_raft_messages`.
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Manual test (end-to-end test)
- No code

Side effects

- Performance regression
    - A little regression on TPS for write scenarios (about 1 ~ 2%).

### Release note 

```release-note
This feature introduces a new mechanism in Raft algorithm which allows a follower to replicate log entries to other followers and learners. This feature is mainly aimed to reduce cross-AZ (short for Availability Zones) traffic in data replication between TiKV instances and TiFlash instances.
```

### CODE REVIEW

**Comment by LintianShi** [comment]
The changes on the raft-rs side is here https://github.com/tikv/raft-rs/pull/485.

File: components/raftstore/src/store/peer.rs [hunk 1]
```diff
@@ -1855,16 +1869,132 @@ where
...
+            }
+
+            // Filter unnecessary MsgAppend and build raft messages.
+            for (pos, msg) in msgs.into_iter().enumerate() {
+                if !skip[pos] && let Some(m) = self.build_raft_message(msg, ctx.self_disk_usage) {
+                    raft_msgs.push(m);
+                }
+            }
+        } else {
+            for msg in msgs {
```
Comment (tonyxuqqi) [inline components/raftstore/src/store/peer.rs:1994]: minor: maybe it's better to put else to if as it's more common code path. Just for better readability.

File: components/raftstore/src/store/peer.rs [hunk 2]
```diff
@@ -1855,16 +1869,132 @@ where
...
+                if msg.get_msg_type() == MessageType::MsgAppend
+                    && self.raft_group.raft.is_replicating(msg.get_to())
+                {
+                    if let Some(to_peer) = self.get_peer_from_cache(msg.get_to()) {
+                        let to_peer_store_id = to_peer.get_store_id();
+                        // Leader and to_peer is not in the same zone.
+                        if let Some(to_peer_zone) = ctx.zone_info.get(&to_peer_store_id) && to_peer_zone != leader_zone
+                        {
+                            // If batch_append is disabled, there may be several MsgAppend to a same peer in msgs.
+                            // The agent should not forward more than one MsgAppend to a single peer.
```
Comment (tonyxuqqi) [inline components/raftstore/src/store/peer.rs:1964]: > // The agent should not forward more than one MsgAppend to a single peer.

Could you elaborate the reason?

File: components/raftstore/src/store/peer.rs [hunk 3]
```diff
@@ -1855,16 +1869,132 @@ where
...
-        msgs: Vec<eraftpb::Message>,
+        mut msgs: Vec<eraftpb::Message>,
     ) -> Vec<RaftMessage> {
         let mut raft_msgs = Vec::with_capacity(msgs.len());
-        for msg in msgs {
-            if let Some(m) = self.build_raft_message(msg, ctx.self_disk_usage) {
-                raft_msgs.push(m);
+        let leader_store_id = self.peer.get_store_id();
+
+        if self.raft_group.raft.follower_replication() && self.is_leader() && let Some(leader_zone) = ctx.zone_info.get(&leader_store_id) {
```
Comment (tonyxuqqi) [inline components/raftstore/src/store/peer.rs:1940]: Could you summarize the algorithm of the following code block?
